File: evals/tasks/observe_yc_startup.ts
import { EvalFunction } from "@/types/evals";

export const observe_yc_startup: EvalFunction = async ({
  debugUrl,
  sessionUrl,
  v3,
  logger,
}) => {
  try {
    const page = v3.context.pages()[0];
    await page.goto("https://www.ycombinator.com/companies");
    await page.waitForLoadState("networkidle");

    const observations = await v3.observe({
      instruction:
        "Click the container element that holds links to each of the startup companies. The companies each have a name, a description, and a link to their website.",
    });

    if (observations.length === 0) {
      return {
        _success: false,
        observations,
        debugUrl,
        sessionUrl,
        logs: logger.getLogs(),
      };
    }

    const possibleLocators = [
      `div._rightCol_i9oky_592`,
      `div._section_i9oky_163._results_i9oky_343`,
    ];

    let foundMatch = false;
    let matchedLocator: string | null = null;

    const nodesEqual = async (
      obsSel: string,
      candSel: string,
    ): Promise<boolean> => {
      return page.evaluate(
        ({ obsSel, candSel }) => {
          function resolve(sel: string): Element | null {
            if (!sel) return null;
            const raw = sel.trim();
            const isXPath =
              /^xpath=/i.test(raw) ||
              raw.startsWith("/") ||
              raw.startsWith("(");
            if (isXPath) {
              try {
                const xp = raw.replace(/^xpath=/i, "");
                return document.evaluate(
                  xp,
                  document,
                  null,
                  XPathResult.FIRST_ORDERED_NODE_TYPE,
                  null,
                ).singleNodeValue as Element | null;
              } catch {
                return null;
              }
            }
            try {
              return document.querySelector(raw);
            } catch {
              return null;
            }
          }
          const a = resolve(obsSel);
          const b = resolve(candSel);
          return a === b;
        },
        { obsSel, candSel },
      );
    };

    for (const observation of observations) {
      try {
        for (const locatorStr of possibleLocators) {
          const isSameNode = await nodesEqual(observation.selector, locatorStr);
          if (isSameNode) {
            foundMatch = true;
            matchedLocator = locatorStr;
            break;
          }
        }

        if (foundMatch) {
          break;
        }
      } catch (error) {
        console.warn(
          `Failed to check observation with selector ${observation.selector}:`,
          error.message,
        );
        continue;
      }
    }

    return {
      _success: foundMatch,
      matchedLocator,
      observations,
      debugUrl,
      sessionUrl,
      logs: logger.getLogs(),
    };
  } catch (error) {
    return {
      _success: false,
      error: error,
      debugUrl,
      sessionUrl,
      logs: logger.getLogs(),
    };
  } finally {
    await v3.close();
  }
};

File: packages/sdk-ruby/evals/tasks/observe/observe_yc_startup.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/observe/observe_yc_startup.ts

Evals.define_task("observe_yc_startup") do |t|
  t.page.goto("https://www.ycombinator.com/companies", wait_until: "networkidle")

  observations = t.stagehand.observe(
    "Click the container element that holds links to each of the startup companies. " \
    "The companies each have a name, a description, and a link to their website.",
  ).data

  if observations.empty?
    next { _success: false, observations: observations.map(&:to_wire) }
  end

  possible_locators = [
    "div._rightCol_18olp_594",
    "div._section_18olp_165._results_18olp_345",
  ]

  # v3 compares backendNodeIds; the v4 Locator exposes no node identity
  # so the same element-identity check is
  # re-expressed in-page: resolve the observed selector and each
  # candidate selector and compare element references.
  matching_selector = lambda do |observed_selector, candidate_selectors|
    t.page.evaluate(<<~JS)
      (() => {
        const observedSelector = #{observed_selector.to_json};
        const candidateSelectors = #{candidate_selectors.to_json};
        const resolve = (selector) => {
          const raw = selector.startsWith("xpath=") ? selector.slice("xpath=".length) : selector;
          if (raw.startsWith("/") || raw.startsWith("(")) {
            const result = document.evaluate(raw, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null);
            return result.singleNodeValue;
          }
          return document.querySelector(raw);
        };
        const observed = resolve(observedSelector);
        if (!observed) return null;
        for (const candidate of candidateSelectors) {
          if (resolve(candidate) === observed) return candidate;
        }
        return null;
      })()
    JS
  end

  found_match = false
  matched_locator = nil

  observations.each do |observation|
    matched = matching_selector.call(observation.selector, possible_locators)
    if matched
      found_match = true
      matched_locator = matched
      break
    end
  rescue StandardError => e
    t.logger.log("Failed to check observation with selector #{observation.selector}: #{e.message}")
    next
  end

  {
    _success: found_match,
    matchedLocator: matched_locator,
    observations: observations.map(&:to_wire),
  }
end
